pkg/volume/volume_linux.go

```diff
@@ -36,60 +39,171 @@ const (
 // SetVolumeOwnership modifies the given volume to be owned by
 // fsGroup, and sets SetGid so that newly created files are owned by
 // fsGroup. If fsGroup is nil nothing is done.
-func SetVolumeOwnership(mounter Mounter, fsGroup *int64) error {
-
+func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy) error {
 	if fsGroup == nil {
 		return nil
 	}
 
+	if skipPermissionChange(mounter, fsGroup, fsGroupChangePolicy) {
+		klog.V(3).Infof("skipping permission and ownership change for volume %s", mounter.GetPath())
+		return nil
+	}
+
 	klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
 
-	return filepath.Walk(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
-		if err != nil {
+	rootDir := mounter.GetPath()
+	// We are not using filepath.Walk() because we do not want to sort files found in a directory before changing
+	// permissions for performance reasons.
+	names, err := readDirNames(rootDir)
+	if err != nil {
+		return err
+	}
+	for _, name := range names {
+		filename := filepath.Join(rootDir, name)
+		err1 := changeAllFilesPermissions(filename, fsGroup, mounter.GetAttributes().ReadOnly)
+		if err1 != nil {
 			return err
 		}
+	}
+	_, err = changeFilePermission(rootDir, fsGroup, mounter.GetAttributes().ReadOnly)
+	return err
+}
 
-		// chown and chmod pass through to the underlying file for symlinks.
-		// Symlinks have a mode of 777 but this really doesn't mean anything.
-		// The permissions of the underlying file are what matter.
-		// However, if one reads the mode of a symlink then chmods the symlink
-		// with that mode, it changes the mode of the underlying file, overridden
-		// the defaultMode and permissions initialized by the volume plugin, which
-		// is not what we want; thus, we skip chown/chmod for symlinks.
-		if info.Mode()&os.ModeSymlink != 0 {
-			return nil
-		}
+func changeAllFilesPermissions(path string, fsGroup *int64, readonly bool) error {
+	fsInfo, err := changeFilePermission(path, fsGroup, readonly)
```

Comment:
The "top" level should be done last right?

Reply:
The top level, i.e the root directory does get done last. It is not "truly" bottoms-up traversal though because I figured that implementing a true DFS will either require keeping track of visited nodes or will require very tight recursive loop and both are bad